Improve XCM Docs (#4533)
w3f/polkadot-wiki: docs/learn/learn-xcm.md

slug: ../learn-xcm
---

> The Cross-Consensus Message Format, or **XCM**, is a **messaging format** and language used to
> define messages between consensus systems.

Reviewer: Does XCM define the messages or does it provide guidelines on how to construct them such that the communication can happen between different consensus systems? Just added an edit that made sense to me. Feel free to rephrase or improve on introducing it. I think the previous author took that sentence from here https://polkadot.network/blog/xcm-the-cross-consensus-message-format#-goals

```suggestion
The Cross-Consensus Message Format, or **XCM**, is a **messaging format** and language used to
communicate between consensus systems.
```

Author: It technically does both, as it *defines* the format in which messages are constructed and executed. However in this context,  `communicate` works better 👍

> One of {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }}'s main initiatives is
> interoperability amongst parachains and any other participating consensus-driven systems. XCM is the
> language in which complex, cross-consensus conversations can occur. Two blockchains can "speak" XCM
> to seamlessly interact with each other using a standard messaging format.

Reviewer: ```suggestion
One of {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }}'s main functionalities is
interoperability amongst parachains and any other participating consensus-driven systems. XCM is the
language through which complex, cross-consensus interactions can occur. Two blockchains can "speak" XCM
to seamlessly interact with each other using a standard messaging format.
```

![xcm](../assets/cross-consensus/xcm.drawio.svg)

:::info

> Remember, a consensus system here means any system or protocol that derives consensus to agree on
> the latest and correct state, whether it's a Polkadot parachain, an EVM smart contract, or other
> bridged consensus systems. We usually discuss XCM in the context of connecting parachains, but
> please bear this in mind!

Reviewer: ```suggestion
We typically discuss XCM in the context of parachains, but please bear this in mind that it expands to the domain of all consensus systems!
Remember, a consensus system here means any system or protocol that achieves finality to agree on
the latest and correct state, whether it's a Polkadot parachain, an EVM smart contract, or other
bridged consensus systems. 
```

:::

XCM is not meant to be only specific to
{{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }}, but rather its primary intention is
to define a **generic** and **common** format amongst different consensus systems to communicate.

> It's important to note that XCM is not how messages are delivered but rather how they look, act, and

Reviewer: ```suggestion
It's important to note that XCM does not define how messages are delivered but rather define how they should look, act, and
```

> contain relative instructions to whatever on-chain actions the message intends to perform.

Reviewer: ```suggestion
contain relative instructions to the on-chain actions the message intends to perform.
```

[**XCMP**](#xcmp-cross-chain-message-passing), or Cross Chain Message Passing, is the actual

> networking-layer protocol to deliver XCM-formatted messages to other participating parachains. There

Reviewer: ```suggestion
network-layer protocol to deliver XCM-formatted messages to other participating parachains. There
```

are also other ways to define transport layer protocols for delivering XCM messages.

> XCM has four, high-level core design principles in which it stands to follow:

Reviewer: ```suggestion
XCM has four high-level core design principles which it stands to follow:
```

1. **Asynchronous**: XCM messages in no way assume that the sender will be blocking on its
   completion.
2. **Absolute**: XCM messages are guaranteed to be delivered and interpreted accurately, in order
   and in a timely fashion. Once a message is sent, one can be sure it will be processed as it was
   intended to be.

> 3. **Asymmetric**: XCM messages, by default, do not have results that let the sender know that the
>    message was received - they are 'fire and forget'. Any results must be separately communicated to
>    the sender with an additional message back to the origin.

Reviewer: ```suggestion
3. **Asymmetric**: XCM messages, by default, do not have results that let the sender know that the
   message was received - they follow the 'fire and forget' paradigm. Any results must be separately communicated to
   the sender with an additional message back to the origin.
```

> 4. **Agnostic**: XCM makes no assumptions about the nature of the Consensus Systems between which

Reviewer: ```suggestion
4. **Agnostic**: XCM makes no assumptions about the nature of the consensus systems between which
```

   the messages are being passed. XCM as a message format should be usable in any system that

>    derives consensus.

Reviewer: ```suggestion
   derives finality through consensus.
```

> These four crucial design decisions allow for XCM messages to be a reliable, yet convenient way to
> properly convey the intentions from one consensus system to another without any compatibility
> limitations.

Reviewer: ```suggestion
These four crucial design decisions allow for XCM messages to be a reliable yet convenient way to
properly convey the intentions from one consensus system to another without any compatibility
issues.
```

:::

> Furthermore, it's essential to realize that XCM themselves messages are _not_ transactions. XCM
> describes how to change the state of the target network, but the message itself doesn't perform the
> state change.

Reviewer: ```suggestion
Furthermore, it's essential to realize that XCM messages by themselves are _not_ considered transactions. XCM
describes how to change the state of the target network, but the message by itself doesn't perform the
state change.
```

> This partly ties what is called **asynchronous composability**, which allows XCM messages to bypass
> the concept of time-constrained mechanisms, like on-chain scheduling, and execute over time in the
> correct order in which it was intended.

Reviewer: ```suggestion
This partly ties to what is called **asynchronous composability**, which allows XCM messages to bypass
the concept of time-constrained mechanisms, like on-chain scheduling and execution over time in the
correct order in which it was intended.
```

3. **Bridging** - introduces the concept of a universal location, which allows for a base reference
   for global consensus systems for multi-hop setups. This location is above the parent relay chain
   or other consensus systems like Ethereum or Bitcoin.

### Why does XCM v3 matter?

A core part of the vision that XCM v3 provides is improving communication between the chains to make
**system parachains** a reality. For example, the Polkadot relay chain handles more than just
parachain management and shared security - it handles user balances/assets, auctions, governance,
and staking. Ideally, the relay chain should be for what it's intended to be - a place for shared
security. System parachains can alleviate these core responsibilities from the relay chain but only
by using a standard format like XCM.

This is where system parachains come in, where each of these core responsibilities can be delegated
to a system parachain respectively.

:::info

XCM v3's bridging, functional multichain decomposition, and programmability upgrades are crucial to
bringing ecosystems together using a common communication abstraction.

:::

### XCM v3 Instruction Overview

> To reflect the functionality above, XCM v3 introduced many new instructions to include within its

Reviewer: ```suggestion
To reflect the functionality above, XCM v3 introduced a multitude of new instructions to include within its
```

messages. This list isn't exhaustive, however contains the necessary instruction sets to showcase
the previously explained concepts of XCM v3.

:::note

> Note most of these instruction definitions came from the source code, of which you may
> [explore for more depth!](https://github.com/paritytech/polkadot/blob/master/xcm/src/v3/mod.rs)

Reviewer: ```suggestion
Note that most of these instruction definitions came from the source code, which you may
[explore for more depth!](https://github.com/paritytech/polkadot/blob/master/xcm/src/v3/mod.rs)
```
